### PR TITLE
chore(eslint): add `safe-await-separator` and tighten rules

### PR DESCRIPTION
**Pull request by michaelfig**
refs: #6745

## Description

Use new-and-improved Jessie await rules, and tighten up the `.eslintrc.cjs` a bit.

For comparison: these new rules now cause about 120 warnings that have since been manually fixed (with a well-placed `await null`) as opposed to the 450+ eslint-disable suppressions of the old `@jessie.js/no-nested-await` that have now been removed.

Also disable `no-await-in-loop` rule for our repository, as our safe-await-separator rule helps protect against accidental misuse.

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -4,18 +4,18 @@ const process = require('process');
 
 const lintTypes = !!process.env.AGORIC_ESLINT_TYPES;
 
-const notLoanDeprecated = [
```
Comment (turadg) [inline .eslintrc.cjs:7]: Why rename this? The intent was to be the complement of "loan" in allDeprecated

File: .eslintrc.cjs [hunk 2]
```diff
@@ -55,6 +55,21 @@ module.exports = {
...
+    // can be trivially satisfied by inserting a non-nested `await null` in an
+    // appropriate place to ensure the rest of the async function runs in a
+    // future turn.  “sometimes synchronous" is a bug farm for particularly
+    // pernicious bugs in which you can combine two correct pieces of code to
+    // have emergent incorrect behavior.  It’s absolutely critical for shared
+    // service code. That means contracts, but it also means kernel components
+    // that are used by multiple clients. So we enable it throughout the repo
+    // and aim for no exceptions.
+    //
+    // TODO the default is 'warn', but upgrade this to 'error' when possible
```
Comment (turadg) [inline .eslintrc.cjs]: Not a blocker but mind making a ticket for this? Then it can be referenced here.

**Review by turadg — COMMENTED** [review]
Core change LGTM! I have a concern about the config cleanup.

File: .eslintrc.cjs (hunk 1 above)
Comment (michaelfig) [reply .eslintrc.cjs:7]: Yeah, that's what I thought it should be when I introduced `notLoanDeprecated`, but I found it weird to have an override targeting the loan contract and saying to use a `notLoan` rule.  I thought it would make more sense to have `loanContractDeprecated`, and then a superset of that for `allDeprecated`.

Maybe a better name would be `deprecatedForLoanContract`?

File: .eslintrc.cjs (hunk 2 above)
Comment (michaelfig) [reply .eslintrc.cjs]: Actually, I upgraded to error and fixed the remaining problems.

File: .eslintrc.cjs (hunk 1 above)
Comment (turadg) [reply .eslintrc.cjs:7]: `deprecatedForLoanContract` doesn't sound like the subset it is. but this is fine. IMO the best solution is to move the sample code somewhere else and that's out of scope.

File: packages/SwingSet/src/controller/controller.js [hunk 3]
```diff
@@ -168,7 +168,7 @@ export async function makeSwingsetController(
   writeSlogObject({ type: 'kernel-init-start' });
 
   writeSlogObject({ type: 'bundle-kernel-start' });
-  // eslint-disable-next-line @jessie.js/no-nested-await
+  await null;
```
Comment (turadg) [inline packages/SwingSet/src/controller/controller.js:171]: took me a minute to see the conditional this is making safe.

File: packages/SwingSet/src/controller/initializeSwingset.js [hunk 4]
```diff
@@ -172,8 +173,8 @@ export function loadBasedir(basedir, options = {}) {
  *    determined.
  */
 async function resolveSpecFromConfig(referrer, specPath) {
+  await null;
```
Comment (turadg) [inline packages/SwingSet/src/controller/initializeSwingset.js:176]: why is this a problem,
```js
return new URL(await resolveModuleSpecifier(specPath, referrer)).pathname;
```
and this is not?
```js
  const url = await resolveModuleSpecifier(specPath, referrer);
  return new URL(url).pathname;
```

File: packages/SwingSet/src/controller/initializeSwingset.js [hunk 5]
```diff
@@ -233,15 +234,14 @@ async function normalizeConfigDescriptor(desc, referrer, expectParameters) {
  *    invalid.
  */
 export async function loadSwingsetConfigFile(configPath) {
+  await null;
```
Comment (turadg) [inline packages/SwingSet/src/controller/initializeSwingset.js:237]: why does the rule error with the `try {}`? The first await is unconditional

File: packages/SwingSet/src/controller/initializeSwingset.js (hunk 4 above)
Comment (erights) [reply packages/SwingSet/src/controller/initializeSwingset.js:176]: Visibility and reviewability of the `await` boundary. If the separating `await` is on an argument expression:

```js
const foo = bar(f(), await g(), h());
```

First and most important, it is too easy to miss. We could require a noisy comment, but it is easier and more pleasant to just require that the `await` expression be extracted to top level of the function body.

Second: This notation makes it somewhat harder to understand what effects happen before the await (`f()` and `g()`) and what after the await (`h()` and `bar(...)`). The fact that the `await` happens *between* `g()` and `h()` is especially surprising and easy to miss if you're reading code in a hurry.

File: packages/SwingSet/src/controller/initializeSwingset.js (hunk 5 above)
Comment (erights) [reply packages/SwingSet/src/controller/initializeSwingset.js:237]:
```js
try {
   await f();
} finally {
   g();
}
```
If `f()` throws, then `g()` happens in the same turn. If `f()` does not throw, then `g()` happens in the next turn.

File: .eslintrc.cjs [hunk 6]
```diff
@@ -55,6 +55,21 @@ module.exports = {
...
+    // appropriate place to ensure the rest of the async function runs in a
+    // future turn.  “sometimes synchronous" is a bug farm for particularly
+    // pernicious bugs in which you can combine two correct pieces of code to
+    // have emergent incorrect behavior.  It’s absolutely critical for shared
+    // service code. That means contracts, but it also means kernel components
+    // that are used by multiple clients. So we enable it throughout the repo
+    // and aim for no exceptions.
+    //
+    // The default is 'warn', but we want to enforce 'error'.
+    '@jessie.js/safe-await-separator': 'error',
```
Comment (erights) [inline .eslintrc.cjs:71]: So the Jessie default remains `warn` and this enforcement of `error` applies only to agoric-sdk? Should we instead make the Jessie default also be `error`?

File: packages/SwingSet/misc-tools/extract-xs-snapshot.js [hunk 7]
```diff
@@ -53,6 +53,7 @@ if (!vatIDToExtract) {
   const { snapPos, hash } = info;
   const snapshot = snapStore.loadSnapshot(vatIDToExtract);
   const fn = `${vatIDToExtract}-${snapPos}-${hash}.xss`;
-  await fs.promises.writeFile(fn, snapshot);
-  console.log(`wrote snapshot to ${fn}`);
+  fs.promises
+    .writeFile(fn, snapshot)
+    .then(() => console.log(`wrote snapshot to ${fn}`));
```
Comment (erights) [inline packages/SwingSet/misc-tools/extract-xs-snapshot.js]:
```suggestion
  E.when(
    fs.promises.writeFile(fn, snapshot),
    () => console.log(`wrote snapshot to ${fn}`));
```
Even when `then` is safe, we should still prefer `E.when` to `then` because of diagnostic info (track-turns deep stacks).

File: .eslintrc.cjs (hunk 6 above)
Comment (michaelfig) [reply .eslintrc.cjs:71]: The `@jessie.js/recommended` rules are the ones that understand `// @jessie-check` in all JS files, and also add warnings for `@jessie.js/safe-await-separator` in all JS files.  When you're actually writing Jessie code, the much stricter `@jessie.js/no-nested-await` is enabled as an `error` rule.

I couldn't bring myself to create an entirely new eslint-plugin just for this rule, and I justified the coupling with Jessie by making it a recommended warning (gently prompting users of the eslint plugin to upgrade it to `error` or downgrade it to `off`).

File: packages/SwingSet/src/controller/initializeSwingset.js (hunk 5 above)
Comment (erights) [reply packages/SwingSet/src/controller/initializeSwingset.js:237]: Even easier to miss:
```js
try {
  await f();
} catch (...) {
  ...
}
g();
```

File: packages/SwingSet/misc-tools/extract-xs-snapshot.js (hunk 7 above)
Comment (michaelfig) [reply packages/SwingSet/misc-tools/extract-xs-snapshot.js]: This file is a one-off script that doesn't have any concurrent activity.  If you think it's still worth the effort, I'll make the change.

File: packages/agoric-cli/src/bin-agops.js [hunk 8]
```diff
@@ -74,13 +73,11 @@ program.addCommand(makeAuctionCommand(logger, { ...procIO, fetch }));
 program.addCommand(makeInterCommand(procIO, { fetch }));
 program.addCommand(makeTestCommand(procIO, { fetch }));
 
-try {
-  await program.parseAsync(process.argv);
-} catch (err) {
+program.parseAsync(process.argv).catch(err => {
```
Comment (erights) [inline packages/agoric-cli/src/bin-agops.js]:
```suggestion
E.when(
  program.parseAsync(process.argv),
  undefined,
  err => {
```
Makes me wonder if we should have an `E.catch` :/

File: packages/SwingSet/src/controller/controller.js [hunk 9]
```diff
@@ -473,8 +473,8 @@ export async function buildVatController(
   };
   const initializationOptions = { verbose, kernelBundles };
   let bootstrapResult;
+  await null;
```
Comment (erights) [inline packages/SwingSet/src/controller/controller.js:476]: Just curious: Why `null`?

No change suggested.

File: packages/cosmic-swingset/scripts/clean-core-eval.js [hunk 10]
```diff
@@ -61,11 +61,15 @@ export const main = async (argv, { readFile, stdout }) => {
...
+    .then(fsp =>
+      main([...process.argv], {
+        readFile: fsp.readFile,
+        stdout: process.stdout,
+      }),
+    )
+    .catch(err => {
+      process.exitCode = 1;
+      console.error(err);
+    });
```
Comment (erights) [inline packages/cosmic-swingset/scripts/clean-core-eval.js]:
```suggestion
  E.when(
    import('fs/promises'),
    fsp =>
      main([...process.argv], {
        readFile: fsp.readFile,
        stdout: process.stdout,
      }),
    ),
    err => {
      process.exitCode = 1;
      console.error(err);
    });
```
Btw, I do realize that I'm making these suggestions on where you've already improved the code to use `.then` or `.catch` rather than `await`. And that an `await` itself also fails to leave behind any diagnostic info (track-turns, deep stacks). At some point we should explore whether there are any debugging / tracing hooks we should ask machines (e.g. XS) to provide. Likely, whatever those are, it would also overcome this problem with `.then`/`.catch`.

File: packages/SwingSet/src/controller/controller.js (hunk 9 above)
Comment (michaelfig) [reply packages/SwingSet/src/controller/controller.js:476]: I'm not sure of the lore behind this idiom, but I see it in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await among other places on the web.

File: packages/inter-protocol/src/price/fluxAggregatorContract.js [hunk 11]
```diff
@@ -131,10 +131,8 @@ export const prepare = async (zcf, privateArgs, baggage) => {
       addr,
       [invitation],
     );
-    if (highPrioritySendersManager) {
-      // eslint-disable-next-line @jessie.js/no-nested-await -- after another await
-      await E(highPrioritySendersManager).add(description, addr);
-    }
+    await (highPrioritySendersManager &&
+      E(highPrioritySendersManager).add(description, addr));
```
Comment (erights) [inline packages/inter-protocol/src/price/fluxAggregatorContract.js:135]: Nice! (and below)

File: packages/inter-protocol/src/reserve/assetReserve.js [hunk 12]
```diff
@@ -66,11 +66,11 @@ export const prepare = async (zcf, privateArgs, baggage) => {
   };
   trace('awaiting takeFeeMint');
   const feeMint = await takeFeeMint();
+  const storageNode = await privateArgs.storageNode;
```
Comment (erights) [inline packages/inter-protocol/src/reserve/assetReserve.js:69]: I think this is an improvement regardless, so no change suggested. But I do want to double check my understanding. Given the `await` already present on line 68, the original code was safe by our new `await` rules, and this is only a readability improvement, yes?

File: packages/inter-protocol/test/psm/test-psm.js [hunk 13]
```diff
@@ -480,19 +480,14 @@ test('mix of trades: failures do not prevent later service', async t => {
...
       case 'want':
-        // eslint-disable-next-line no-await-in-loop
         await wantMinted(ix, give, want, ok, wants);
         break;
       default:
         assert.fail(kind);
     }
-    if (kind === 'give') {
-      // eslint-disable-next-line no-await-in-loop
-    }
```
Comment (erights) [inline packages/inter-protocol/test/psm/test-psm.js:495]: What was the original code trying to do? Regardless, it was a noop, so removing is correctness preserving. No change suggested.

File: packages/inter-protocol/src/reserve/assetReserve.js (hunk 12 above)
Comment (michaelfig) [reply packages/inter-protocol/src/reserve/assetReserve.js:69]: This is a Jessie contract, so it uses `@jessie.js/no-nested-await`.

File: packages/swingset-liveslots/test/test-vpid-liveslots.js [hunk 14]
```diff
@@ -293,7 +293,6 @@ async function doVatResolveCase23(t, which, mode, stalls) {
         // another few turns. We wait some number of turns before using p1
         // again, to exercise as many race conditions as possible.
         for (let i = 0; i < stalls; i += 1) {
-          // eslint-disable-next-line no-await-in-loop
           await Promise.resolve();
```
Comment (erights) [inline packages/swingset-liveslots/test/test-vpid-liveslots.js:296]: Where is the boundary `await` that makes this one safe? Did your new lint rule not flag this one?

Comment (michaelfig) [reply packages/swingset-liveslots/test/test-vpid-liveslots.js:296]: It's in a test, all of which are explicitly exempt from the await rules.

**Review by erights — CHANGES_REQUESTED** [review]
I was reviewing this under the assumption that your new lint rule was catching any code that violated the new `await` rule. But then I saw `test-vpid-liveslots.js`, which I commented on, which makes me wonder whether the lint rule sometimes misses.

File: packages/wallet/api/src/lib-wallet.js [hunk 15]
```diff
@@ -1895,6 +1897,7 @@ export function makeWalletRoot({
   const importBankAssets = async bank => {
     observeIteration(subscribeEach(E(bank).getAssetSubscription()), {
```
Comment (erights) [inline packages/wallet/api/src/lib-wallet.js]:
```suggestion
    observeIteration(subscribeEach(E(bank).getAssetSubscription()), harden({
```
or
```suggestion
    observeIteration(subscribeEach(E(bank).getAssetSubscription()), Far(..., {
```
Nothing to do with this PR, but a good drive-by fix. I noticed this because its absence made the code harder to visually parse, which is interesting. In context, on first visual pass, I mistook this open curly to be a function-body open curly rather than an object-literal open curly.

**Review by erights — COMMENTED** [review]
I did not review the lint implementation of the new rule.

I did not review whether the inserted `await null;`s were correctness preserving. Given the number of them, I do not know how to do so in an economical manner. Did you find any economical heuristic when you decided to insert them?

I did assume that
   * the implementation of the new rule is correct, which `test-vpid-liveslots.js` now makes me suspicious of
   * that all the code now passes that correct lint rule.

Aside from these caveats, all the rest LGTM

File: .eslintrc.cjs (hunk 6 above)
Comment (erights) [reply .eslintrc.cjs:71]: I do think it is worth making both of these changes to Jessie. 
   * IMO the `@jessie.js/no-nested-await` should default to 'off', but would like your thoughts.
   * ***IF*** `@jessie.js/no-nested-await` is downgraded from `error`, then `@jessie.js/safe-await-separator` should be upgraded to `error`. ELSE no change suggested.

But since Jessie is a separate repo, this PR does not need to wait on that Jessie change.

File: packages/SwingSet/misc-tools/extract-xs-snapshot.js (hunk 7 above)
Comment (erights) [reply packages/SwingSet/misc-tools/extract-xs-snapshot.js]: Lots of code is written by imitation of other code. I do think this effort is worth it since this PR introduces the `.then` rather than just preserving one. Thanks.

File: packages/SwingSet/src/controller/controller.js (hunk 9 above)
Comment (erights) [reply packages/SwingSet/src/controller/controller.js:476]: Unaware of those precedents. Good to know, thanks.

**Comment by michaelfig** [comment]
> I did not review whether the inserted `await null;`s were correctness preserving. Given the number of them, I do not know how to do so in an economical manner. Did you find any economical heuristic when you decided to insert them?

I tried to study and discern the intent of each `await` that failed the rule.  In basically every case, there wasn't any controversy (i.e. no prepare-commit patterns that would be disrupted by a non-nested `await`), then I was looking for things that seemed suspicious.  That went a lot faster, because the vast majority was not suspicious.

> I did assume that
> 
> * the implementation of the new rule is correct, which `test-vpid-liveslots.js` now makes me suspicious of

I'm pretty sure the rule is correct.  It's just that `.eslintrc.cjs` explicitly disables the await lint rules for test files (you can see the exempt filename patterns there if you are not satisfied with this answer).

> Aside from these caveats, all the rest LGTM

I've made the requested changes.  Would you consider approving?

**Review by erights — APPROVED** [review]
LGTM, thanks!
